### PR TITLE
Pbi 1799

### PR DESCRIPTION
**Pull request by aeng7**
Conviva plugin
-Currently tracks: play, pause, buffering, stop, ad start, ad end
-TODO: SSAI integration, bitrate information, other Conviva Content Metadata

### CODE REVIEW

File: js/plugins/conviva.js [hunk 1]
```diff
@@ -0,0 +1,962 @@
...
+ * @classdesc Conviva SDK plugin that works with the Ooyala Analytics Framework.
+ * @param {object} framework The Analytics Framework instance
+ */
+var ConvivaAnalyticsPlugin = function(framework)
+{
+  var _framework = framework;
+  var name = "conviva";
+  var version = "v1";
+  var id;
+  var _active = true;
```
Comment (dchauOoyala) [inline js/plugins/conviva.js]: What is the significance here of prefixing "active" with the underscore?

Comment (aeng7) [inline js/plugins/conviva.js]: Was pulled from the analytics template. Removed from all the plugins as it was not used.

File: js/plugins/conviva.js [hunk 2]
```diff
@@ -0,0 +1,941 @@
...
+      convivaClient.releasePlayerStateManager(playerStateManager);
+    }
+  };
+
+  /**
+   * Gathers all relevant application information for a particular video playback and stores this
+   * information inside a Conviva ContentMetadata object. Clears the last session if necessary. After
+   * the ContentMetadata is created, the Conviva Client will create a session for tracking.
+   * @private
+   * @method ConvivaAnalyticsPlugin#tryBuildConvivaContentMetadata
```
Comment (pilievOoyala) [inline js/plugins/conviva.js]: should we return a boolean if it succeeded? A suggestion, not a requirement.

File: js/plugins/conviva.js [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+   * @public
+   * @method ConvivaAnalyticsPlugin#setMetadata
+   * @param  {object} metadata The metadata for this plugin
+   */
+  this.setMetadata = function(metadata)
+  {
+    if (validateConvivaMetadata(metadata)) {
+      convivaMetadata = metadata;
+      trySetupConviva();
+    }
```
Comment (pilievOoyala) [inline js/plugins/conviva.js]: should we destroy the plugin if it's not valid?

File: js/plugins/conviva.js [hunk 4]
```diff
@@ -0,0 +1,941 @@
...
+    if (systemFactory)
+    {
+      systemFactory.release();
+      systemFactory = null;
+    }
+  };
+
+  var validSession = function()
+  {
+    return currentConvivaSessionKey !== Conviva.Client.NO_SESSION_KEY && currentConvivaSessionKey !== null;
```
Comment (pilievOoyala) [inline js/plugins/conviva.js]: what about undefined instead of just null check?

File: js/plugins/conviva.js [hunk 5]
```diff
@@ -0,0 +1,941 @@
...
+   */
+  var trackAdStart = function()
+  {
+    if (canTrack())
+    {
+      var adPosition = null;
+      if (currentPlayhead <= 0)
+      {
+        OO.log("[Conviva-Ooyala] Playing preroll");
+        adPosition = Conviva.Client.AdPosition.PREROLL;
```
Comment (pilievOoyala) [inline js/plugins/conviva.js]: are we guaranteed Conviva.Client exists? Probably should have some null checks.

File: test/unit-tests/testConviva.js [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+      "gatewayUrl":"testUrl",
+      "customerKey":"testKey"
+    };
+    plugin.setMetadata(metadata);
+    return plugin;
+  };
+
+  it('Test Conviva Plugin Validity', function() {
+    var convivaPluginFactory = require(SRC_ROOT + "plugins/conviva.js");
+    expect(convivaPluginFactory).not.toBeNull();
```
Comment (pilievOoyala) [inline test/unit-tests/testConviva.js]: shoudl this be a check for defined? Can it return undefined?

File: test/unit-tests/testConviva.js [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+    var plugin = new convivaPluginFactory(framework);
+    expect(framework.validatePlugin(plugin)).toBe(true);
```
Comment (pilievOoyala) [inline test/unit-tests/testConviva.js]: can you also do negative testing with a bad plugin. Make sure that works as well.

File: test/unit-tests/testConviva.js [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+    simulator.simulateAdBreakEnded();
+    expect(Conviva.currentPlayerStateManager.currentPlayerState).toBe(Conviva.PlayerStateManager.PlayerState.STOPPED);
+    expect(Conviva.currentClient.adPlaying).toBe(false);
+    expect(Conviva.currentClient.adEndSessionId).toBe(sessionId);
+  });
+  
+  //new sessions: replay, embed code change, etc
+  it('Conviva Plugin can start new session on replay',function()
+  {
+    var plugin = createPlugin(framework);
```
Comment (pilievOoyala) [inline test/unit-tests/testConviva.js]: to be safe, can you add in a test that validates doing replay twice.  That way if you going into a bad state on replay, the 3 play through will catch it.  same for the next test where it changes the embed code.

File: test/unit-tests/testConviva.js [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+    simulator.simulatePlayerLoad({
+      embedCode: "testEmbedCode",
+      title: "testTitle",
+      duration: 60000
+    });
+
+    simulator.simulateAdBreakStarted();
+    simulator.simulateAdPlayback({
+      adType: OO.Analytics.AD_TYPE.LINEAR_VIDEO,
+      adMetadata: {
```
Comment (dchauOoyala) [inline test/unit-tests/testConviva.js]: brace should be on the newline here. same comment for a few more places below

File: test/unit-tests/testConviva.js [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    expect(Conviva.currentClient.adPlaying).toBe(false);
+
+    simulator.simulateVideoProgress({
+      playheads: [10]
+    });
+
+    simulator.simulateAdBreakStarted();
+    simulator.simulateAdPlayback({
+      adType: OO.Analytics.AD_TYPE.LINEAR_VIDEO,
+      adMetadata: {
```
Comment (dchauOoyala) [inline test/unit-tests/testConviva.js]: newline brace

File: test/unit-tests/testConviva.js [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+
+    simulator.simulateContentComplete({
+      streamPosition: 60
+    });
+    expect(Conviva.currentClient.adPlaying).toBe(false);
+
+    simulator.simulateAdBreakStarted();
+    simulator.simulateAdPlayback({
+      adType: OO.Analytics.AD_TYPE.LINEAR_VIDEO,
+      adMetadata: {
```
Comment (dchauOoyala) [inline test/unit-tests/testConviva.js]: newline brace

File: js/plugins/conviva.js [hunk 12]
```diff
@@ -0,0 +1,941 @@
...
+
+  /**
+   * [Required Function] Set the metadata for this plugin.
+   * @public
+   * @method ConvivaAnalyticsPlugin#setMetadata
+   * @param  {object} metadata The metadata for this plugin
+   */
+  this.setMetadata = function(metadata)
+  {
+    if (validateConvivaMetadata(metadata)) {
```
Comment (dchauOoyala) [inline js/plugins/conviva.js]: newline brace :[

File: js/plugins/conviva.js [hunk 13]
```diff
@@ -0,0 +1,941 @@
...
+   * If the required metadata and Conviva SDK is available, this function will
+   * create the system interface and settings to create an instance of the Conviva
+   * Client object. Will also try to create the Conviva Content Metadata once the
+   * Client object is available.
+   * @private
+   * @method ConvivaAnalyticsPlugin#trySetupConviva
+   */
+  var trySetupConviva = function()
+  {
+    if (convivaMetadata && sdkLoaded()){
```
Comment (dchauOoyala) [inline js/plugins/conviva.js]: newline brace :/

File: js/plugins/conviva.js (hunk 5 above)
Comment (aeng7) [inline js/plugins/conviva.js]: The SDK should guarantee that Conviva.Client exists. We check if the SDK has loaded. This should be sufficient.

File: test/unit-tests/testConviva.js (hunk 7 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: I'll create a bad plugin file and test it in another branch.

**Comment by aeng7** [comment]
Note that the coverage numbers are lower than expected because of the system interface functions that are required by the SDK. I pulled the ones from Conviva's sample app as suggested for a baseline HTML5 implementation.

File: js/plugins/conviva.js (hunk 12 above)
Comment (aeng7) [inline js/plugins/conviva.js]: : {

File: js/plugins/conviva.js (hunk 13 above)
Comment (aeng7) [inline js/plugins/conviva.js]: : {

File: test/unit-tests/testConviva.js (hunk 11 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: We've agreed to keep braces on the same line for object definitions.

File: test/unit-tests/testConviva.js (hunk 10 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: We've agreed to keep braces on the same line for object definitions.

File: test/unit-tests/testConviva.js (hunk 9 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: We've agreed to keep braces on the same line for object definitions.

File: test/unit-tests/testConviva.js (hunk 6 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: done

File: test/unit-tests/testConviva.js (hunk 8 above)
Comment (aeng7) [inline test/unit-tests/testConviva.js]: done

File: js/plugins/conviva.js (hunk 2 above)
Comment (aeng7) [inline js/plugins/conviva.js]: done

**Pull request merged**